text/0022-gas-spot-market.md

```diff
@@ -0,0 +1,215 @@
+
+### Chainspec parameters
+
+There are several natural parameters that govern the process we outlined. While it would reduce the scope of proposed work to leave these hardcoded, the optimal values cannot be determined in advance and will change as the platform grows, making it imperative that these parameters be included in the chainspec.
+
+#### General gas limit
+
+The limit on spot gas reservations by general deploys.
+
+#### WASM-less transfer dedicated gas limit
+
+The limit on spot gas reservations of WASM-less deploys, naturally stated as a multiple of desired number of such deploys multiplied by their (fixed) gas reservation. 
+
+### Deploy channels
+
+We have already mentioned the concept of a dedicated gas limit for WASM-less transfers in our description of the proposed chainspec parameters. To avoid crowding out WASM-less transfers, which are expected to be critical for platform health and public perception, there should be a separate "lane" for such transfers and this parameter specifies the width of this "lane." Selection among WASM-less transfers would work the same as it does for general deploys, but with sorting by WTP alone, since the gas reservation is fixed for these. Moreover, WASM-less transfers that are not selected for inclusion would be eligible for inclusion as general deploys, with stated WTP and fixed gas reservation.
+
+### Lifecycle of a deploy
+
+#### Construction
+
+(user-side)
+
+1. User creates a deploy in any supported manner, additionally specifying willingness to pay and gas reservation.
+    
+    a.  If the deploy is a WASM-less transfer, the gas reservation is fixed.
+
+#### Submission
+
+(user-side)
+
+1. User submits the deploy to any node as usual.
+
+#### Ordering
+
+(validator-side)
+
+1. Block proposer orders available deploys for inclusion in a protoblock.
+
+    a. Sort by WTP (greatest to least).
+    
+    b. Sort by gas reservation (greatest to least).
+
+#### Balance check & allocation
+
+(validator-side, proposal)
+
+For each channel, starting with WASM-less transfers, the following process is to be followed:
+
+1. Traverse the ordered list of deploys, checking balances.
+    
+    a. Check that originating account purse holds an amount equal to willingness to pay times gas reservation, record the remainder for this originating account purse.
+    
+    b. If sufficient tokens, include the deploy in the protoblock.
+    
+    c. Upon encountering the same originating account purse subsequently, check willingness to pay times gas reservation against remainder, and update the remainder.
+    
+    d. If sufficient tokens in the remainder, include the deploy in the protoblock.
+    
+2. Continue traversing deploy-by-deploy until the gas limit is full.
+
+3. Once gas limit cannot be filled by the next deploy with an eligible balance or remainder, skip deploys until a deploy small enough is encountered to be considered, otherwise continuing as above.
+
```

Comment:
I remain confused whether we are _absolutely_ decided on FIFO. The small complexities in this CEP can be removed. The actual substance is just a single sort, instead of grabbing deploys from the buffer as they come (however, for full block gas utilization, we probably do want to skip "too large" deploys _even_ _in_ _FIFO,_ instead of just stopping and sending out a potentially near empty block).